seed/seedwriter/seed20.go

```diff
@@ -0,0 +1,316 @@
+}
+
+func (pol *policy20) checkSnapChannel(ch channel.Channel, whichSnap string) error {
+	// TODO: consider allowing some channel overrides for >=signed
+	// Core 20 models?
+	return pol.checkAllowedDangerous()
+}
+
+func (pol *policy20) systemSnap() *asserts.ModelSnap {
+	return makeSystemSnap("snapd")
+}
+
+func (pol *policy20) modelSnapDefaultChannel() string {
+	// We will use latest/stable as default, model that want something else
+	// will need to to speficy a default-channel
+	return "latest/stable"
+}
+
+func (pol *policy20) extraSnapDefaultChannel() string {
+	// We will use latest/stable as default
+	// TODO: consider using just "stable" for these?
+	return "latest/stable"
+}
+
+func (pol *policy20) checkBase(info *snap.Info, availableSnaps *naming.SnapSet) error {
+	base := info.Base
+	if base == "" {
+		if info.GetType() != snap.TypeGadget && info.GetType() != snap.TypeApp {
+			return nil
+		}
+		base = "core"
+	}
+
+	if availableSnaps.Contains(naming.Snap(base)) {
+		return nil
+	}
+
+	whichBase := fmt.Sprintf("its base %q", base)
+	if base == "core16" {
+		if availableSnaps.Contains(naming.Snap("core")) {
+			return nil
+		}
+		whichBase += ` (or "core")`
+	}
+
+	return fmt.Errorf("cannot add snap %q without also adding %s explicitly", info.SnapName(), whichBase)
+}
+
+func (pol *policy20) needsImplicitSnaps(*naming.SnapSet) (bool, error) {
+	// no implicit snaps with Core 20
+	// TODO: unless we want to support them for extra snaps
+	return false, nil
+}
+
+func (pol *policy20) implicitSnaps(*naming.SnapSet) []*asserts.ModelSnap {
+	return nil
+}
+
+func (pol *policy20) implicitExtraSnaps(*naming.SnapSet) []*OptionsSnap {
+	return nil
+}
+
+type tree20 struct {
+	opts *Options
+
+	snapsDirPath string
+	systemDir    string
+
+	systemSnapsDirEnsured bool
+}
+
+func (tr *tree20) mkFixedDirs() error {
+	tr.snapsDirPath = filepath.Join(tr.opts.SeedDir, "snaps")
+	tr.systemDir = filepath.Join(tr.opts.SeedDir, "systems", tr.opts.Label)
+
+	if err := os.MkdirAll(tr.snapsDirPath, 0755); err != nil {
+		return err
+	}
+
+	return os.MkdirAll(tr.systemDir, 0755)
+}
+
+func (tr *tree20) ensureSystemSnapsDir() error {
+	if tr.systemSnapsDirEnsured {
+		return nil
+	}
+	snapsDir := filepath.Join(tr.systemDir, "snaps")
+	if err := os.MkdirAll(snapsDir, 0755); err != nil {
+		return err
+	}
+	tr.systemSnapsDirEnsured = true
+	return nil
+}
+
+func (tr *tree20) snapsDir() string {
+	// XXX what about extra snaps?
+	return tr.snapsDirPath
+}
+
+func (tr *tree20) localSnapPath(*SeedSnap) string {
+	panic("XXX dangerous features not implemented yet")
+}
+
+func (tr *tree20) writeAssertions(db asserts.RODatabase, modelRefs []*asserts.Ref, snapsFromModel []*SeedSnap, extraSnaps []*SeedSnap) error {
+	assertsDir := filepath.Join(tr.systemDir, "assertions")
+	if err := os.MkdirAll(assertsDir, 0755); err != nil {
+		return err
+	}
+
+	writeRefs := func(fname string, refsGen func(stop <-chan struct{}) <-chan *asserts.Ref) error {
```

Comment:
This is cool stuff - while reading it I could not help wondering if there is a way to make it a bit less indirect but I can't come up with anything so its fine.